### PR TITLE
Proposal for a new UserInterface

### LINKED ISSUE #24: Make UserInterface::getIdentity() return an array
I've been having issues with the UserInterface. For the current application I'd like to have the ID, email address and name available. However for some other applications I only want a username. So in other words, the userinterface is pretty limited. 

If we change the getIdentity method to an array or maybe add an extra getSessionAttributes() method to return an array with extra attributes, it would make it a lot more flexible.

```php
public function getSessionAttributes() : array
{
    return [
        'id' => $this->id,
        'name' => $this->name,
        'email' => $this->email,
        'username' => $this->username,
    ];
}
```

### PR DESCRIPTION
**Pull request by ezimuel**
This PR proposes a new [UserInterface](https://github.com/ezimuel/zend-expressive-authentication/blob/feature/new-user-interface/src/UserInterface.php) and a [DefaultUser](https://github.com/ezimuel/zend-expressive-authentication/blob/feature/new-user-interface/src/DefaultUser.php) implementation.

**UPDATE** see the update commits and comment [here](https://github.com/zendframework/zend-expressive-authentication/pull/28#issuecomment-390711855).

This PR fixes #24 and it is a replacement of #25.

**NOTE:** this PR contains BC break. It also requires to update the `zend-expressive-authentication` adapters like [zend-expressive-authentication-oauth2](https://github.com/zendframework/zend-expressive-authentication-oauth2),  [zend-expressive-authentication-session](https://github.com/zendframework/zend-expressive-authentication-session), [zend-expressive-authentication-zendauthentication](https://github.com/zendframework/zend-expressive-authentication-zendauthentication), [zend-expressive-authentication-basic](https://github.com/zendframework/zend-expressive-authentication-basic).

### CODE REVIEW

File: src/UserInterface.php [hunk 1]
```diff
@@ -11,15 +11,25 @@
...
+     */
+    public function setRoles(array $roles) : void;
+
     /**
      * Get all user roles
      *
      * @return string[]
      */
-    public function getUserRoles() : array;
+    public function getRoles() : array;
```
Comment (geerteltink) [inline src/UserInterface.php:24]: 👍

**Review by geerteltink — COMMENTED** [review]
This is a step in the right direction, but it doesn't fix #24 and it's not really a replacement for #25. This is actually what I had in an early stage in #25. Basically the problem to store custom data in the session still exists. Only the identity and the user roles are stored. There is no method in the user interface the session persistence classes can rely on to (re)store additional data.

The idea in #24 and #25 is to store additional data to the identity and roles. Let's say I have a User object and want to persist it in a session with this data:
```
[
    'identity' => 'me@example.com',
    'name' => 'xtreamwayz',
    'roles' => [],
    'claims' => [], 
]
```
I don't see how zend-expressive-authentication-session knows how to (re)store the additional data with the changes in this PR.

I like the changes in this PR and I think it is a step in the right direction, but I think it doesn't go far enough.

File: docs/book/v1/intro.md [hunk 2]
```diff
@@ -42,6 +55,33 @@ if a user has been authenticated or not, e.g. it can be used to verify the
 authorization level of a user (for this scope, it is consumed by
 [zend-expressive-authorization](https://github.com/zendframework/zend-expressive-authorization)).
 
+## Default User class
+
+We provided a default user class, implemented by `Zend\Authentication\DefaultUser`.
+This class is a basic implementation of `UserInterface`. This default class
```
Comment (weierophinney) [inline docs/book/v1/intro.md]: Replace the first sentence with:

```markdown
We provide a default implementation of `UserInterface` via the class `Zend\Expressive\Authentication\DefaultUser`.
```

File: docs/book/v1/intro.md [hunk 3]
```diff
@@ -23,17 +23,30 @@ interface UserInterface
...
-     *
-     * @return string
      */
-    public function getIdentity(): string;
+    public function getIdentity() : string;
+
+    /**
+     * Set the user identity
+     */
+    public function setIdentity(string $identity) : void
```
Comment (weierophinney) [inline docs/book/v1/intro.md]: I'm a little skeptical of adding setter methods to the `UserInterface`. The reason is because this could potentially lead to cases where middleware resets the identify, roles, etc. — which is fine for altering the user instance passed _deeper_ into the request layer, but could have ramifications for _parent_ layers as well. I think we should prefer immutability here.

File: src/ConfigProvider.php [hunk 4]
```diff
@@ -58,11 +59,15 @@ public function getDependencies() : array
...
                 // UserRepositoryInterface::class => UserRepository\Htpasswd::class
+                UserInterface::class => DefaultUser::class
             ],
             'factories' => [
                 AuthenticationMiddleware::class => AuthenticationMiddlewareFactory::class,
                 UserRepository\Htpasswd::class => UserRepository\HtpasswdFactory::class,
                 UserRepository\PdoDatabase::class => UserRepository\PdoDatabaseFactory::class
+            ],
+            'invokables' => [
+                DefaultUser::class => DefaultUser::class
```
Comment (weierophinney) [inline src/ConfigProvider.php]: What you and I had discussed earlier was that the user "service" would map to a _factory class_ that the authentication middleware would then compose; by doing that, we both de-couple how the instance is produced as well as the type produced.

I wonder if that might answer some of the questions from @xtreamwayz , as individual adapters could provide their own user implementations **and** the factories for producing them.

File: src/DefaultUser.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ * @see       https://github.com/zendframework/zend-expressive-authentication for the canonical source repository
+ * @copyright Copyright (c) 2017 Zend Technologies USA Inc. (https://www.zend.com)
+ * @license   https://github.com/zendframework/zend-expressive-authentication/blob/master/LICENSE.md New BSD License
+ */
+
+declare(strict_types=1);
+
+namespace Zend\Expressive\Authentication;
+
+class DefaultUser implements UserInterface
```
Comment (weierophinney) [inline src/DefaultUser.php]: I would make this class `final`, and mark the properties as `private`. The interface is pretty slim, and it's less likely that a developer will extend this class than provide a clean implementation.

**Comment by ezimuel** [comment]
@xtreamwayz I think the need to store user information in PHP Session using `zend-expressive-authentication-session` is specific for this adapter. This PR enables the usage of custom `UserInterface` implementation in `zend-expressive-authentication`. I just have a look at the session implementation and maybe we can just store the `$user` instance in the session ([here](https://github.com/zendframework/zend-expressive-authentication-session/blob/master/src/PhpSession.php#L89-L92)). PHP serialize/unserialize object values automatically in the session.

File: src/UserRepository/Htpasswd.php [hunk 6]
```diff
@@ -61,10 +65,12 @@ public function authenticate(string $credential, string $password = null) : ?Use
...
 
-        return $found
-            && password_verify($password === null ? '' : $password, $hash)
-                ? $this->generateUser($credential)
-                : null;
+        if ($found && password_verify($password === null ? '' : $password, $hash)) {
+            $this->user->setIdentity($credential);
+            $this->user->setRoles($this->getRolesFromUser($credential));
+            return $this->user;
+        }
```
Comment (weierophinney) [inline src/UserRepository/Htpasswd.php]: This here demonstrates why passing a factory to the repository is a better approach. If we use the same instance, we cannot have concurrent requests (ala Swoole), as the instance is in the container, and can be changed by any process running at any time.

If a factory were composed in this class instead, this would become something like:

```php
return $this->userFactory->generate($credential, $this->getRolesFromUser($credential));
```

Which would guarantee that the user generated is _unique to this request_.

**Review by weierophinney — CHANGES_REQUESTED** [review]
I've noted a few changes, some of which we discussed elsewhere earlier today. The main points are that I think users should be immutable, and that I think we should be composing a _factory_ and not an _instance_ for the user within repositories. This allows generating the user per-request, and prevents issues of concurrency and unexpected state changes.

Additionally, I think we may need a way to store arbitrary user-related data in the `UserInterface` instances. This would be separate from the identity and roles, but could vary based on adapter. Doing so would allow us to persist that information between sessions more easily, and also ensure we can communicate that information to deeper middleware layers that may need access to it.

File: src/UserRepository/Htpasswd.php [hunk 7]
```diff
@@ -61,10 +65,12 @@ public function authenticate(string $credential, string $password = null) : ?Use
         }
         fclose($handle);
 
-        return $found
-            && password_verify($password === null ? '' : $password, $hash)
-                ? $this->generateUser($credential)
-                : null;
+        if ($found && password_verify($password === null ? '' : $password, $hash)) {
```
Comment (michalbundyra) [inline src/UserRepository/Htpasswd.php]: We can have here `$password ?? ''` instead of `$password === null ? '' : $password`.

File: src/DefaultUser.php [hunk 8]
```diff
@@ -0,0 +1,37 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive-authentication for the canonical source repository
+ * @copyright Copyright (c) 2017 Zend Technologies USA Inc. (https://www.zend.com)
```
Comment (michalbundyra) [inline src/DefaultUser.php]: 2018

File: src/ConfigProvider.php [hunk 9]
```diff
@@ -58,11 +59,15 @@ public function getDependencies() : array
                 // AuthenticationInterface::class => Basic\BasicAccess::class,
                 // Provide an alias for the UserRepository adapter based on your application needs.
                 // UserRepositoryInterface::class => UserRepository\Htpasswd::class
+                UserInterface::class => DefaultUser::class
```
Comment (michalbundyra) [inline src/ConfigProvider.php]: Trailing comma after the last element in the array, here and in many other places.

**Comment by geerteltink** [comment]
> Additionally, I think we may need a way to store arbitrary user-related data in the UserInterface instances. This would be separate from the identity and roles, but could vary based on adapter. Doing so would allow us to persist that information between sessions more easily, and also ensure we can communicate that information to deeper middleware layers that may need access to it.

Have a look at this UserInterface: https://github.com/zendframework/zend-expressive-authentication/pull/25/files#diff-6e0a82375e1ea981d662cd8109eee2db

It solves a few concerns mentioned here. It make the class immutable up to a certain level and doesn't use setters. Besides that it provides a way to persist arbitrary user-related data between sessions.

File: src/UserRepository/Htpasswd.php (hunk 6 above)
Comment (geerteltink) [reply src/UserRepository/Htpasswd.php]: I don't think a factory is the right solution. You still need to pass the identity and roles. What about extra data, how would you pass that without knowing what data it is? And also how do you get that data again so you can persist it in a session?

File: src/ConfigProvider.php (hunk 4 above)
Comment (geerteltink) [reply src/ConfigProvider.php]: It's a step closer to a solution. However, it's not only about a User object. It's also about storing in and retrieving from a session.

**Comment by ezimuel** [comment]
@weierophinney I updated the PR with the following changes:
- `DefaultUser` now is a final class with immutable values;
- I provided a `UserInterfaceFactory` class that generates a `DefaultUser` instance using `$identity` and `$roles`;
- I remove the alias in the ConfigProvider;
- I updated the documentation;

@xtreamwayz I think the `UserInterface` should simply remain as it is. If you want to provide a custom implementation we can use a different `UserInterfaceFactory` implementation. Regarding the Session data, we can serialize the user's object (actually this is done automatically by PHP Session). What do you think?

File: docs/book/v1/intro.md [hunk 10]
```diff
@@ -42,6 +40,37 @@ if a user has been authenticated or not, e.g. it can be used to verify the
 authorization level of a user (for this scope, it is consumed by
 [zend-expressive-authorization](https://github.com/zendframework/zend-expressive-authorization)).
 
+## Default User class
+
+We provide a default implementation of `UserInterface` via the class `Zend\Expressive\Authentication\DefaultUser`.
+This class is a basic implementation of `UserInterface`. This class is final and
+it's immutable, that means you cannot change it state at runtime.
```
Comment (weierophinney) [inline docs/book/v1/intro.md]: Remove the sentence "This class is a basic implementation..."; it's redundant.

Rephrase the next sentence to: "The class is final and immutable, in order to prevent runtime changes."

File: docs/book/v1/intro.md [hunk 11]
```diff
@@ -42,6 +40,37 @@ if a user has been authenticated or not, e.g. it can be used to verify the
...
+In order to set the identity and the user's role we provided a default factory
+class that generates a `UserInterface` object. This factory is
+`Zend\Expressive\Authentication\UserInterfaceFactory`. This class uses a `generate`
```
Comment (weierophinney) [inline docs/book/v1/intro.md]: Rephrase to:

```markdown
We provide a factory class for generating `DefaultUser` instances via `Zend\Expressive\Authentication\UserFactory`.
```

Note: I think the name of the factory should either be `UserFactory` or `DefaultUserFactory`, and not contain the word `Interface`.

Additionally, if we will expect a `generate()` method, we likely need to provide an interface to describe the factory, to ensure all implementations are written the same.

File: docs/book/v1/intro.md [hunk 12]
```diff
@@ -42,6 +40,37 @@ if a user has been authenticated or not, e.g. it can be used to verify the
...
+function to create a `UserInterface` instance passing the identity and the roles
+(if any) of the user.
+
+If you want, you can customize the `UserInterfaceFactory` using your custom
+`UserInterface` implementation. You need to change the service configuration as
+follows:
```
Comment (weierophinney) [inline docs/book/v1/intro.md]: I'd rephrase to:

```markdown
You may provide custom `UserInterface` implementations by providing your own user factory implementation. Once you have created one, change the service configuration as follows:
```

File: docs/book/v1/intro.md [hunk 13]
```diff
@@ -42,6 +40,37 @@ if a user has been authenticated or not, e.g. it can be used to verify the
...
+
+```php
+// src/ConfigProvider.php
```
Comment (weierophinney) [inline docs/book/v1/intro.md]: I would demonstrate this as straight configuration, and not as part of a `ConfigProvider`. The reasons:

- Application developers will likely want to know how to override the configuration at the application level, which is in the `config/autoload/*.php` files.
- Package developers (those creating adapters) will know that the configuration goes in a `ConfigProvider` class.

File: docs/book/v1/intro.md [hunk 14]
```diff
@@ -75,26 +104,18 @@ Basic Access Authentication* adapter and the *htpasswd* file as the user
...
 
 ```php
-// ConfigProvider.php
-
-use Zend\Expressive\Authentication\AuthenticationInterface;
-use Zend\Expressive\Authentication\UserRepositoryInterface;
-
-class ConfigProvider
-{
+// src/ConfigProvider.php
```
Comment (weierophinney) [inline docs/book/v1/intro.md]: Same comment here; do this as a dependency configuration array, not part of a `ConfigProvider`.

File: src/UserInterfaceFactory.php [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+ * @license   https://github.com/zendframework/zend-expressive-authentication/blob/master/LICENSE.md New BSD License
+ */
+
+declare(strict_types=1);
+
+namespace Zend\Expressive\Authentication;
+
+use Psr\Container\ContainerInterface;
+
+class UserInterfaceFactory
```
Comment (weierophinney) [inline src/UserInterfaceFactory.php]: As noted previously: drop the `Interface` verbiage, and potentially add a `Default` prefix.

File: src/UserRepository/Htpasswd.php [hunk 16]
```diff
@@ -20,17 +21,20 @@
...
+    /**
+     * @var UserInterfaceFactory
+     */
+    private $userFactory;
+
     /**
      * @throws Exception\InvalidConfigException
      */
-    public function __construct(string $filename)
+    public function __construct(string $filename, UserInterfaceFactory $userFactory)
```
Comment (weierophinney) [inline src/UserRepository/Htpasswd.php]: Right here is a good example of why the factory should likely have an interface (e.g., `UserFactoryInterface`. That way, developers do not need to _extend_ the factory class, but can instead provide their own implementation of the interface.

**Review by weierophinney — CHANGES_REQUESTED** [review]
I've noted some additional changes.

In relation to the comments from @xtreamwayz —  should the `UserInterface` allow access to other user-related data? So, for instance:

```php
$fullName = $user->getDetail('full_name', $user->getIdentity());
$email = $user->getDetail('email', '');
$userDetails = $user->getDetails();
```

Such changes would require a third, optional argument to the user factory method to accommodate, but could answer:

- How consumers can predictably retrieve additional user details beyond the identity and roles.
- How providers can populate those details.

Thoughts?

File: docs/book/v1/intro.md (hunk 13 above)
Comment (ezimuel) [reply docs/book/v1/intro.md]: Ok

File: docs/book/v1/intro.md (hunk 12 above)
Comment (ezimuel) [reply docs/book/v1/intro.md]: Ok

File: docs/book/v1/intro.md (hunk 11 above)
Comment (ezimuel) [reply docs/book/v1/intro.md]: I'm rewriting the factory using `__invoke()`, so we don't need to specify another interface.

File: docs/book/v1/intro.md (hunk 10 above)
Comment (ezimuel) [reply docs/book/v1/intro.md]: Ok

File: docs/book/v1/intro.md (hunk 14 above)
Comment (ezimuel) [reply docs/book/v1/intro.md]: Ok

File: src/UserInterfaceFactory.php (hunk 15 above)
Comment (ezimuel) [reply src/UserInterfaceFactory.php]: Changed in `DefaultUserFactory`.

File: src/UserRepository/Htpasswd.php (hunk 16 above)
Comment (ezimuel) [reply src/UserRepository/Htpasswd.php]: I changed it with a `callable`.

**Comment by ezimuel** [comment]
@weierophinney, @xtreamwayz  I did all the proposed changes. I changed the `UserInterface` adding `getDetails()` and `getDetail(string $name, $default = null)` to be able to manage additional user information. I also added the `getUserDetails(string $identity)` function in `UserRepositoryInterface`. An implementation of this method is available in `UserRepository\PdoDatabase`, [here](https://github.com/ezimuel/zend-expressive-authentication/blob/feature/new-user-interface/src/UserRepository/PdoDatabase.php#L127-L163).

File: src/UserRepositoryInterface.php [hunk 17]
```diff
@@ -26,5 +26,13 @@ public function authenticate(string $credential, string $password = null) : ?Use
      * @param string $identity
      * @return string[]
      */
-    public function getRolesFromUser(string $identity) : array;
+    public function getUserRoles(string $identity) : array;
```
Comment (geerteltink) [inline src/UserRepositoryInterface.php]: As asked in #29, why is this in the interface? This could be used internally in a private function in `PdoDatabase`. All the interface should care about is authenticate a user and return a UserInterface object.

File: src/UserRepositoryInterface.php [hunk 18]
```diff
@@ -26,5 +26,13 @@ public function authenticate(string $credential, string $password = null) : ?Use
...
+
+    /**
+     * Get the user details if present.
+     *
+     * @param string $identity
+     * @return string[]
+     */
+    public function getUserDetails(string $identity) : array;
```
Comment (geerteltink) [inline src/UserRepositoryInterface.php]: Same as above, see #29

**Review by geerteltink — COMMENTED** [review]
It's not entirely what I had in mind but it's a nice compromise and definitely will give users the ability to use extra data.

File: src/UserRepositoryInterface.php (hunk 17 above)
Comment (ezimuel) [reply src/UserRepositoryInterface.php]: @xtreamwayz actually you right! I'm refactoring the PR removing these functions. Thanks!

**Comment by ezimuel** [comment]
@xtreamwayz I removed the functions from `UserRepositoryInterface`, leaving only the following method:

```php
public function authenticate(string $credential, string $password = null) : ?UserInterface;
```

Moreover, I added more unit tests for `PdoDatabase` and updated the documentation with the new interfaces.

@weierophinney I think the PR is now completed.

File: src/UserRepository/Htpasswd.php [hunk 19]
```diff
@@ -39,6 +42,7 @@ public function __construct(string $filename)
             ));
         }
         $this->filename = $filename;
+        $this->userFactory = $userFactory;
```
Comment (weierophinney) [inline src/UserRepository/Htpasswd.php]: One thing we did with the `ResponseFactory` was to wrap this in an additional callable to provide type safety:

```php
$this->userFactory = function (string $identity, array $roles = [], array $details = []) : UserInterface use ($userFactory) {
    return $userFactory($identity, $roles, $details);
};
```

This ensures that if the callable does not return the type expected, it will raise a `TypeError`. 

I can do that change on merge.

**Review by weierophinney — APPROVED** [review]
